Add Document Icon Button, AutoLayout Quick Settings

saagarjha opened the pull request:
- [X] This change has been discussed with the author (at least, it's been mentioned many times)
- [x] It implements / fixes issue #315, #353.

---

**Description:**
A proper fix for both issues outlined above.  Please test it out, especially with long filenames and the fullscreen transition.

Caveats:
* This PR uses internal APIs. Essentially, `NSWindow` keeps changing the `NSTextField` for the title whenever switching between fullscreen and windowed mode. It also doesn't make the actual text field publicly available. Thus, the two solutions to accessing it are the API I used or iterating through `window.standardWindowButtons(.documentIcon).superview.subviews` for a `NSTextField`. Your choice on which one we use.
* I removed a line in `windowWillExitFullScreen` for reverting the appearance; I don't see what its purpose is. Please take a look to see if it's important.
* There is are fundamental issues with the layering of the views in `MainWindowController` that will require a lot of work to fix. I've left it alone for now, but we will need to take a look at it sometime.

saagarjha commented:
A picture is worth a thousand words, so here's 4000:

The pop-up button:
<img width="687" alt="screen shot 2017-02-14 at 10 47 29 am" src="https://cloud.githubusercontent.com/assets/13786931/22944589/038b9a0a-f2a6-11e6-9a53-c58abda04f92.png">
<img width="730" alt="screen shot 2017-02-14 at 10 47 58 am" src="https://cloud.githubusercontent.com/assets/13786931/22944586/034eb086-f2a6-11e6-9175-398b7455747a.png">

The title automatically truncates:
<img width="689" alt="screen shot 2017-02-14 at 10 53 15 am" src="https://cloud.githubusercontent.com/assets/13786931/22944588/035d90a6-f2a6-11e6-98a3-e7500781eef2.png">
<img width="851" alt="screen shot 2017-02-14 at 10 53 28 am" src="https://cloud.githubusercontent.com/assets/13786931/22944587/0356ffde-f2a6-11e6-8e2b-de95b76a3197.png">

saagarjha commented:
@lhc70000 Fixed the merge conflicts.

saagarjha commented:
@lhc70000 Could you please take a look at this, and merge it if appropriate? I've gone through and fixed the merge conflicts.

lhc70000 commented:
Sorry, I was busy with my coursework last week…… I'll take a look later today and promise nothing will be committed to `develop` before merging this PR.

saagarjha commented:
Thanks :)

lhc70000 commented on iina/MainWindowController.swift:
```diff
@@ -162,7 +151,21 @@ class MainWindowController: NSWindowController, NSWindowDelegate {
-  @IBOutlet weak var titleTextField: NSTextField!
+  var standardWindowButtons: [NSButton] {
+    get {
+      return ([.closeButton, .miniaturizeButton, .zoomButton, .documentIconButton] as [NSWindowButton]).flatMap {
```
Why `flatMap`? Isn't `map` enough?

saagarjha replied:
`flatMap` strips out `nil`.

saagarjha commented on iina/QuickSettingViewController.swift:
```diff
@@ -41,6 +38,9 @@ class QuickSettingViewController: NSViewController, NSTableViewDataSource, NSTab
   @IBOutlet weak var subTabBtn: NSButton!
   @IBOutlet weak var tabView: NSTabView!
 
+	@IBOutlet weak var buttonTopConstraint: NSLayoutConstraint!
```
Removed.

saagarjha commented on iina/PlaylistViewController.swift:
```diff
@@ -134,7 +141,7 @@ class PlaylistViewController: NSViewController, NSTableViewDataSource, NSMenuDel
     }
   }
-
+  
```
Removed.

saagarjha commented:
> QuickSettingView:
> 
> The animation is annoying to me (I set auto hide duration to 2s, so it often jumps up and down). It'll distract user's attention. I also asked some user's opinion and they don't like it a lot.

Ok, I removed it.

> There're warnings in QuickSettingView.xib: Frame for "Custom View" will be different at run time.

This appears to be a bug in Interface Builder, since the frames can't be fixed. I'll file a Radar.

> The runtime position for "1x" is also incorrect.

Fixed.

> Please don't change the layout, like sub position slider.

Reverted.

> Titlebar:
> 
> I prefer using window.standardWindowButtons(.documentIcon).superview.subviews.

Ok.

> Please set the titlebar's truncate mode to middle.

The system's behavior is to clip the end; this doesn't appear to be customizable.

> Please don't include irrelevant stuff in one PR (I've told you twice), and do discuss with others when implementing a new feature. I can't merge this PR since the QuickSettingView is totally out of the plan. My proposed solution: close this PR and open another one for titlebar (which only includes changes to MainWindowController), and discuss whether we need to change anything else.

This isn't "irrelevant stuff", and AutoLayout for the QuickSettingView had already been discussed in #253. The document icon button is "above" the QuickSettingView when it wasn't before, so I had to shift the QuickSettingView down to accommodate. This was impossible to do without fixing the layout for QuickSettingView. Unless you want the title bar and QuickSettingView to overlap, you'll need to merge these together.

lhc70000 commented:
0

> The document icon button is "above" the QuickSettingView when it wasn't before

Sorry, I didn't even know this before. This is an important problem and definitely need discussion (you should discuss with me once found it, but not changing UI by yourself).

My thought based on current situation:
- Use fixed padding (i.e. `downShift`) for quick setting view
- Or, add back the animation (with a shorter interval)

For this PR please just revert back to the initial animation version and I can handle the remaining stuff later.

1

> This appears to be a bug in Interface Builder, since the frames can't be fixed

For slider labels and equalizer, why not putting them in a box and only add auto layout for the box? I think it won't affect the overall layout.
The messed up xib makes it hard to do further adjustment on the controls. Is the custom view approach a commonly used trick? 

2

> The system's behavior is to clip the end; this doesn't appear to be customizable.

Now that we got the `NSTextField` itself, there's nothing we can't do on it 😉. Confirmed the following code works:
```swift
if let attrTitle = titleTextField?.attributedStringValue.mutableCopy() as? NSMutableAttributedString {
    let p = attrTitle.attribute(NSParagraphStyleAttributeName, at: 0, effectiveRange: nil) as! NSMutableParagraphStyle
    p.lineBreakMode = .byTruncatingMiddle
    attrTitle.addAttribute(NSParagraphStyleAttributeName, value: p, range: NSRange(location: 0, length: attrTitle.length))
}
```

3

> > Don't change the layout

It also include "Load External Audio" and other buttons.

4

> This isn't "irrelevant stuff"

You should told me detailed info like said above. But besides this, there're more "irrelevant stuff" like fixing font weight warning, `Utility.fatal` etc., they are actually pretty good, but made it difficult to review and I could easily . Separate PR next time, please.

---

Do you have an IM account like telegram, or perhaps a Gitter or Slack chatroom is better? So that we can discuss immediately once something happened, if you plan to make further contribution on this project.

saagarjha commented:
> 0
> 
> The document icon button is "above" the QuickSettingView when it wasn't before
> Sorry, I didn't even know this before. This is an important problem and definitely need discussion (you should discuss with me once found it, but not changing UI by yourself).
> 
> My thought based on current situation:
> 
> Use fixed padding (i.e. downShift) for quick setting view
> Or, add back the animation (with a shorter interval)
> For this PR please just revert back to the initial animation version and I can handle the remaining stuff later.
> 

Ok, I've put the animation back. We can work on improving it later.

> 1
> 
> This appears to be a bug in Interface Builder, since the frames can't be fixed
> For slider labels and equalizer, why not putting them in a box and only add auto layout for the box? I think it won't affect the overall layout.
> The messed up xib makes it hard to do further adjustment on the controls. Is the custom view approach a commonly used trick?
> 

The "normal" method in this case is to let the Stack View handle the layout; however, this method doesn't work on OS X Yosemite. We need to use views in the middle and let them have equal widths to space them out.

> 2
> 
> The system's behavior is to clip the end; this doesn't appear to be customizable.
> Now that we got the NSTextField itself, there's nothing we can't do on it 😉. Confirmed the following code works:
> 
> if let attrTitle = titleTextField?.attributedStringValue.mutableCopy() as? NSMutableAttributedString {
>     let p = attrTitle.attribute(NSParagraphStyleAttributeName, at: 0, effectiveRange: nil) as! NSMutableParagraphStyle
>     p.lineBreakMode = .byTruncatingMiddle
>     attrTitle.addAttribute(NSParagraphStyleAttributeName, value: p, range: NSRange(location: 0, length: attrTitle.length))
> }

Are you sure? This will have to be done repeatedly, whenever the titleTextField changes (which it does often).

> 3
> 
> Don't change the layout
> It also include "Load External Audio" and other buttons.
> 

Reverted.

> 4
> 
> This isn't "irrelevant stuff"
> You should told me detailed info like said above. But besides this, there're more "irrelevant stuff" like fixing font weight warning, Utility.fatal etc., they are actually pretty good, but made it difficult to review and I could easily . Separate PR next time, please.

The changes to `Utility.fatal` came out of the new use of `fatal` from a background thread, which would crash the thread but not the whole app.

> 
> Do you have an IM account like telegram, or perhaps a Gitter or Slack chatroom is better? So that we can discuss immediately once something happened, if you plan to make further contribution on this project.

Does IINA have a Slack or Gitter? I could join that.

lhc70000 commented:
> Are you sure? This will have to be done repeatedly, whenever the titleTextField changes (which it does often).

Seems that we only need to set it once (Cocoa will update the string value only).

> Does IINA have a Slack or Gitter? I could join that.

Created one and invited you 🎉 https://gitter.im/iina-dev/Lobby
